### PR TITLE
Enable macOS cross-compilation

### PR DESCRIPTION
**Pull request by str4d**
Includes code cherry-picked from the following upstream PRs:

- bitcoin/bitcoin#7809
  - The `native_cctools` fix.
- bitcoin/bitcoin#8210
  - The macOS toolchain bump.
- bitcoin/bitcoin#9891
- bitcoin/bitcoin#15581
  - The `tar` change.
- bitcoin/bitcoin#16049
  - The `native_cctools` change.

Build instructions:
- Fetch `MacOSX10.11.sdk` from e.g. https://github.com/phracker/MacOSX-SDKs/releases
- Extract it into `depends/SDKs` (creating that folder first)
- `HOST=x86_64-apple-darwin11 ./zcutil/build.sh`

### CODE REVIEW

**Comment by rex4539** [comment]
Make sure macOS 10.15 Catalina is supported.

Currently, compilation fails in so many ways, I can't even dare to file a bug :)

File: depends/packages/native_cctools.mk [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  rm -f toolchain/lib/libc++abi.so* && \
+  echo "#!/bin/sh" > toolchain/bin/$(host)-dsymutil && \
+  echo "exit 0" >> toolchain/bin/$(host)-dsymutil && \
+  chmod +x toolchain/bin/$(host)-dsymutil && \
+  tar --no-same-owner --strip-components=1 -xf $($(package)_source)
+endef
+
+define $(package)_set_vars
+$(package)_config_opts=--target=$(host) --disable-lto-support
+$(package)_ldflags+=-Wl,-rpath=\\$$$$$$$$\$$$$$$$$ORIGIN/../lib
```
Comment (daira) [inline depends/packages/native_cctools.mk]: Add a comment explaining the escaping.

File: depends/packages/rust.mk [hunk 2]
```diff
@@ -10,10 +10,12 @@ $(package)_sha256_hash_freebsd=20d062493d01f1816014fe9dbe883bda06f1828a6ddbfb7ee
 
 # Mapping from GCC canonical hosts to Rust targets
 # If a mapping is not present, we assume they are identical
+$(package)_rust_target_x86_64-apple-darwin11=x86_64-apple-darwin
```
Comment (daira) [inline depends/packages/rust.mk]: It's this going to need to be continually updated for new macOS versions?

**Comment by daira** [comment]
Re: https://github.com/bitcoin/bitcoin/pull/9891 , an alternative would be to require the fixed version of ld64.

**Comment by defuse** [comment]
When I tried to build it I ran into an error [similar to this one](https://github.com/theos/theos/issues/357), I worked around it by symlinking `/usr/lib/libtinfo.so.5` to `/usr/lib/libtinfo.so.6`. Next I got a bunch of errors, starting with...

```
make[5]: Entering directory '/home/firexware/zcash/zcash/depends/work/build/x86_64-apple-darwin11/native_cctools/807d6fd1be5d2224872e381870c0a75387fe05e6-53854c2fbe5/cctools/ld64/src/ld/parsers'
/bin/sh ../../../../libtool  --tag=CXX   --mode=compile /home/firexware/zcash/zcash/depends/work/build/x86_64-apple-darwin11/native_cctools/807d6fd1be5d2224872e381870c0a75387fe05e6-53854c2fbe5/toolchain/bin/clang++ -DPACKAGE_NAME=\"cctools\" -DPACKAGE_TARNAME=\"cctools\" -DPACKAGE_VERSION=\"877.8\" -DPACKAGE_STRING=\"cctools\ 877.8\" -DPACKAGE_BUGREPORT=\"t.poechtrager@gmail.com\" -DPACKAGE_URL=\"\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DEMULATED_HOST_CPU_TYPE=16777223 -DEMULATED_HOST_CPU_SUBTYPE=3 -D__STDC_LIMIT_MACROS=1 -D__STDC_CONSTANT_MACROS=1 -DHAVE_EXECINFO_H=1 -DHAVE_UUID_UUID_H=1 -I.   -I/home/firexware/zcash/zcash/depends/x86_64-apple-darwin11/native/include     -D__DARWIN_UNIX03 -Wall -Wno-long-long -Wno-import -Wno-format -Wno-deprecated -Wno-unused-variable -Wno-unused-private-field -Wno-unused-function -Wno-invalid-offsetof -Wno-int-conversion -Wno-char-subscripts -Wno-shift-negative-value  -D__LITTLE_ENDIAN__=1 -I../../../../include -I../../../../include/foreign -I../../../../ld64/src/abstraction -I../../../../ld64/src/ld -I../../../../ld64/src/3rd -I../../../../ld64/src -O3 -std=c++0x  -fblocks -c -o libParsers_la-archive_file.lo `test -f 'archive_file.cpp' || echo './'`archive_file.cpp
libtool: compile:  /home/firexware/zcash/zcash/depends/work/build/x86_64-apple-darwin11/native_cctools/807d6fd1be5d2224872e381870c0a75387fe05e6-53854c2fbe5/toolchain/bin/clang++ -DPACKAGE_NAME=\"cctools\" -DPACKAGE_TARNAME=\"cctools\" -DPACKAGE_VERSION=\"877.8\" "-DPACKAGE_STRING=\"cctools 877.8\"" -DPACKAGE_BUGREPORT=\"t.poechtrager@gmail.com\" -DPACKAGE_URL=\"\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DEMULATED_HOST_CPU_TYPE=16777223 -DEMULATED_HOST_CPU_SUBTYPE=3 -D__STDC_LIMIT_MACROS=1 -D__STDC_CONSTANT_MACROS=1 -DHAVE_EXECINFO_H=1 -DHAVE_UUID_UUID_H=1 -I. -I/home/firexware/zcash/zcash/depends/x86_64-apple-darwin11/native/include -D__DARWIN_UNIX03 -Wall -Wno-long-long -Wno-import -Wno-format -Wno-deprecated -Wno-unused-variable -Wno-unused-private-field -Wno-unused-function -Wno-invalid-offsetof -Wno-int-conversion -Wno-char-subscripts -Wno-shift-negative-value -D__LITTLE_ENDIAN__=1 -I../../../../include -I../../../../include/foreign -I../../../../ld64/src/abstraction -I../../../../ld64/src/ld -I../../../../ld64/src/3rd -I../../../../ld64/src -O3 -std=c++0x -fblocks -c archive_file.cpp  -fPIC -DPIC -o .libs/libParsers_la-archive_file.o
In file included from archive_file.cpp:32:
In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/vector:60:
In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_algobase.h:64:
In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_pair.h:59:
In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/move.h:55:
/usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/type_traits:1018:48: error: '_Tp' does not refer to a value
      : public __bool_constant<__is_assignable(_Tp, _Up)>
                                               ^
/usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/type_traits:1016:21: note: declared here
  template<typename _Tp, typename _Up>
                    ^
/usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/type_traits:1018:53: error: '_Up' does not refer to a value
      : public __bool_constant<__is_assignable(_Tp, _Up)>
                                                    ^
/usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/type_traits:1016:35: note: declared here
  template<typename _Tp, typename _Up>
                                  ^
/usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/type_traits:1019:5: error: expected class name
    { };
    ^
/usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/type_traits:132:31: error: no member named 'value' in 'std::is_copy_assignable<std::_Rb_tree_node_base *>'
    : public conditional<_B1::value, _B2, _B1>::type
                         ~~~~~^
/usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_pair.h:379:3: note: in instantiation of template class 'std::__and_<std::is_copy_assignable<std::_Rb_tree_node_base *>,
      std::is_copy_assignable<std::_Rb_tree_node_base *> >' requested here
                __and_<is_copy_assignable<_T1>,
                ^
/usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_tree.h:2086:5: note: in instantiation of template class 'std::pair<std::_Rb_tree_node_base *, std::_Rb_tree_node_base *>' requested here
    _M_get_insert_unique_pos(const key_type& __k)
    ^
In file included from archive_file.cpp:32:
In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/vector:60:
In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_algobase.h:64:
In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_pair.h:59:
In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/move.h:55:
/usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/type_traits:132:31: error: no member named 'value' in 'std::is_move_assignable<std::_Rb_tree_node_base *>'
    : public conditional<_B1::value, _B2, _B1>::type
                         ~~~~~^
...
```

I'm on Arch Linux.

File: depends/packages/rust.mk (hunk 2 above)
Comment (str4d) [reply depends/packages/rust.mk]: We would add additional mappings to the same Rust target.

**Comment by str4d** [comment]
Rebased to fix merge conflict (i.e. update Rust version).

> I'm on Arch Linux.

I don't think we need to support cross-compiling **from** every Linux platform; as long as it can be done from our usual Debian-based platforms then I think this is sufficient.

I confirmed that I can build this branch on Ubuntu 18.04.3 LTS inside WSL. I do get a bunch of ld warnings:
```
[...]
ld: warning: direct access in ___cxx_global_var_init.12 to global weak symbol guard variable for libzcash::IncrementalMerkleTree<32ul, libzcash::PedersenHash>::emptyroots means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in libzcash::IncrementalMerkleTree<29ul, libzcash::SHA256Compress>::empty_root() to global weak symbol libzcash::IncrementalMerkleTree<29ul, libzcash::SHA256Compress>::emptyroots means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in libzcash::IncrementalMerkleTree<32ul, libzcash::PedersenHash>::empty_root() to global weak symbol libzcash::IncrementalMerkleTree<32ul, libzcash::PedersenHash>::emptyroots means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in libzcash::NoteDecryption<585ul>::decrypt(std::__1::array<unsigned char, 601ul> const&, uint256 const&, uint256 const&, unsigned char) const to global weak symbol typeinfo for libzcash::note_decryption_failed means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in libzcash::PaymentDisclosureNoteDecryption<585ul>::decryptWithEsk(std::__1::array<unsigned char, 601ul> const&, uint256 const&, uint256 const&, uint256 const&, unsigned char) const to global weak symbol typeinfo for libzcash::note_decryption_failed means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in uint256 Hash<std::__1::__wrap_iter<char*> >(std::__1::__wrap_iter<char*>, std::__1::__wrap_iter<char*>) to global weak symbol Hash<std::__1::__wrap_iter<char*> >(std::__1::__wrap_iter<char*>, std::__1::__wrap_iter<char*>)::pblank means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in void boost::throw_exception<boost::bad_optional_access>(boost::bad_optional_access const&) to global weak symbol typeinfo for boost::wrapexcept<boost::bad_optional_access> means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::bad_optional_access> >::rethrow() const to global weak symbol typeinfo for boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::bad_optional_access> > means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::detail::system_category_default_error_condition_posix(int) to global weak symbol boost::system::system_category()::system_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::detail::system_category_default_error_condition_posix(int) to global weak symbol boost::system::generic_category()::generic_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::generic_category() to global weak symbol boost::system::generic_category()::generic_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::generic_category() to global weak symbol guard variable for boost::system::generic_category()::generic_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::generic_category() to global weak symbol boost::system::generic_category()::generic_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::generic_category() to global weak symbol guard variable for boost::system::generic_category()::generic_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::generic_category() to global weak symbol guard variable for boost::system::generic_category()::generic_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::system_category() to global weak symbol boost::system::system_category()::system_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::system_category() to global weak symbol guard variable for boost::system::system_category()::system_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::system_category() to global weak symbol boost::system::system_category()::system_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::system_category() to global weak symbol guard variable for boost::system::system_category()::system_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in boost::system::system_category() to global weak symbol guard variable for boost::system::system_category()::system_category_instance means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in LockedPageManager::Instance() to global weak symbol LockedPageManager::CreateInstance() means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
```
but the binaries are built and linked.

**Comment by mdr0id** [comment]
testACK on ARM:

```
marsh@pop-os:~/ARM_zcash/zcash$ file ./src/zcashd
./src/zcashd: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=3b4580e9bf267be574194987973403b50e24ec85, for GNU/Linux 3.7.0, with debug_info, not stripped
marsh@pop-os:~/ARM_zcash/zcash$ file ./src/zcash-cli
./src/zcash-cli: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=2cccb954f5bb909852d7e8c48aa97cf5dd93db81, for GNU/Linux 3.7.0, with debug_info, not stripped
marsh@pop-os:~/ARM_zcash/zcash$ file ./src/zcash-tx
./src/zcash-tx: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=1bd9750ad646e3a6995793aada420c332634a51c, for GNU/Linux 3.7.0, with debug_info, not stripped
marsh@pop-os:~/ARM_zcash/zcash$ file ./src/zcash-gtest
./src/zcash-gtest: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=f27db35c7035d8a69a87a89a575437850fc44b44, for GNU/Linux 3.7.0, with debug_info, not stripped
```

**Comment by mdr0id** [comment]
testACK for MacOS cross compile (from ubuntu 18.04; ubuntu 19.04 failed to cross compile for Mac):

```
mtron@woot:~/ARM_MAC_zcash/zcash/src$ file zcashd
zcashd: Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|WEAK_DEFINES|BINDS_TO_WEAK|PIE|HAS_TLV_DESCRIPTORS>
mtron@woot:~/ARM_MAC_zcash/zcash/src$ file zcash-cli
zcash-cli: Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|WEAK_DEFINES|BINDS_TO_WEAK|PIE>
mtron@woot:~/ARM_MAC_zcash/zcash/src$ file zcash-gtest
zcash-gtest: Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|WEAK_DEFINES|BINDS_TO_WEAK|PIE|HAS_TLV_DESCRIPTORS>
mtron@woot:~/ARM_MAC_zcash/zcash/src$ file zcash-tx
zcash-tx: Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|WEAK_DEFINES|BINDS_TO_WEAK|PIE|HAS_TLV_DESCRIPTORS>
mtron@woot:~/ARM_MAC_zcash/zcash/src$ 
```

**Comment by mdr0id** [comment]
@zkbot r+

**Comment by daira** [comment]
To fix those linker warnings, I think you need to add `-fvisibility=hidden -fvisibility-inlines-hidden` to the compiler (C and C++) flags for Zcash and all dependencies. See https://stackoverflow.com/questions/9894961/strange-warnings-from-the-linker-ld and https://stackoverflow.com/questions/8685045/xcode-with-boost-linkerid-warning-about-visibility-settings .

**Pull request merged**